Reporter point source

lynnmunday opened the pull request:
closes #18528 

This is removing the `VectorPostprocessorPointSource` object from the MOOSE and replacing it with `ReporterPointSource`.  There will not be a deprecation period for `VectorPostprocessorPointSource`.  The `VectorPostprocessorPointSource` was put into MOOSE on 10/2020 and was only being used by the porous flow module.  The porous flow tests have been switched over to the reporter point source.

lynnmunday committed:
Enable reporter data to be read in by vectorPostprocessorPointSource  close #18528

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/791132/) on 5717b60 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18529/site)

This comment will be updated on new commits.

dschwen commented on framework/src/dirackernels/VectorPostprocessorPointSource.C:
```diff
@@ -20,28 +20,46 @@ VectorPostprocessorPointSource::validParams()
-  params.addParam<std::string>("value_name", "value", "name of column containing values.");
+      "The name of the VectorPostprocessor containing positions and corresponding load values.  "
+      "This will be appended onto the reporter name as <vector_postprocessor>/<vector_name>.");
+  params.addParam<std::string>("x_coord_name", "x", "name of reporter containing x coordinates.");
+  params.addParam<std::string>("y_coord_name", "y", "name of reporter containing y coordinates.");
+  params.addParam<std::string>("z_coord_name", "z", "name of reporter containing z coordinates.");
+  params.addParam<std::string>("value_name", "value", "name of reporter containing values.");
```
```suggestion
  params.addParam<ReporterName>("x_coord_name", "x", "name of reporter containing x coordinates.");
  params.addParam<ReporterName>("y_coord_name", "y", "name of reporter containing y coordinates.");
  params.addParam<ReporterName>("z_coord_name", "z", "name of reporter containing z coordinates.");
  params.addParam<ReporterName>("value_name", "value", "name of reporter containing values.");
```

dschwen replied:
Huh, I'm confused. The parameters don't make any sense anymore. The VPP was supposed to have coords and data, and teh next params were just the column names in that VPP, now they are separate reporters?

zachmprince replied:
Reporter names have the syntax `object_name/value_name`. So the parameter `vector_postprocessor` will specify the `object_name` and the `x_coord_name`, `y_coord_name`, etc. specify the `value_name`.

dschwen replied:
Ok, then the doc strings need to be fixed

lynnmunday replied:
The api doesn't change so it will still work as it did before.  The reporter interface makes is written like this so that a consumer can take in different reporters.  It's not helpful in this case but there are examples of using different reporters in the tests.  I think I want to pass around a reporter in the optimization code and I need this vppPointSource to apply the misfit so it would be nice if it could read from a reporter.

lynnmunday replied:
I don't know what you mean by this

> Ok, then the doc strings need to be fixed

dschwen replied:
`x_coord_name` is not "name of reporter", it is the name of the value

lynnmunday replied:
@dschwen or @zachmprince I fixed the doc strings and it passes the tests.  Will you merge it?

lynnmunday replied:
@dschwen will you review this so that it can be merged

lynnmunday replied:
@dschwen will you merge this?

zachmprince commented on framework/src/dirackernels/VectorPostprocessorPointSource.C:
```diff
@@ -20,41 +20,86 @@ VectorPostprocessorPointSource::validParams()
-    _z_coord(getVectorPostprocessorValue(
-        "vector_postprocessor", getParam<std::string>("z_coord_name"), true))
+    ReporterInterface(this),
+    _vpp_values(
+        isParamValid("vector_postprocessor")
+            ? getReporterValueByName<std::vector<Real>>(
+                  getParam<std::string>("vector_postprocessor") + "/" +
+                  getParam<std::string>("value_name"))
+            : getReporterValueByName<std::vector<Real>>(getParam<std::string>("value_name"))),
+    _x_coord(
+        isParamValid("vector_postprocessor")
+            ? getReporterValueByName<std::vector<Real>>(
+                  getParam<std::string>("vector_postprocessor") + "/" +
+                  getParam<std::string>("x_coord_name"))
+            : getReporterValueByName<std::vector<Real>>(getParam<std::string>("x_coord_name"))),
+    _y_coord(
+        isParamValid("vector_postprocessor")
+            ? getReporterValueByName<std::vector<Real>>(
+                  getParam<std::string>("vector_postprocessor") + "/" +
+                  getParam<std::string>("y_coord_name"))
+            : getReporterValueByName<std::vector<Real>>(getParam<std::string>("y_coord_name"))),
+    _z_coord(isParamValid("vector_postprocessor")
+                 ? getReporterValueByName<std::vector<Real>>(
+                       getParam<std::string>("vector_postprocessor") + "/" +
+                       getParam<std::string>("z_coord_name"))
+                 : getReporterValueByName<std::vector<Real>>(getParam<std::string>("z_coord_name")))
```
I don't really like all this duplicate code. I would recommend just having a helper function like this:

```cpp
const std::vector<Real> &
VectorPostprocessorPointSource::getPointDataHelper(const std::string & param)
{
  ReporterName rname = isParamValid("vector_postprocessor") ?
                       ReporterName(getParam<std::string>("vector_postprocessor"), getParam<std::string>(param)) :
                       ReporterName(getParam<std::string>(param));
  return getReporterValueByName<std::vector<Real>>(rname);
}
```

The you can just do this in the constructor:

```cpp
...
_vpp_values(getPointDataHelper("value_name")),
_x_coord(getPointDataHelper("x_coord_name")),
_y_coord(getPointDataHelper("y_coord_name")),
_z_coord(getPointDataHelper("z_coord_name"))
```

zachmprince replied:
Also, as an extra precaution, it might be good to make sure we are receiving all the values on all processors:
```cpp
getReporterValueByName<std::vector<Real>>(rname, REPORTER_MODE_REPLICATED)
```

lynnmunday replied:
Thanks @zachmprince.   I pasted in your changes.

zachmprince reviewed (changes requested):
Looks good! Just small suggestion.

lynnmunday committed:
adding Zachs code review changes

lynnmunday committed:
replaced vppPointSource with ReporterPointSource, changed api, changed documentation, updated all tests to use reporterPointSource

lynnmunday commented:
@zachmprince @dschwen Will you review again?  I changed everything over to use only reporter syntax and I changed all of the tests that used vppPointSource.  Only porous flow was using it so the api change shouldn't be a big deal.

zachmprince commented on framework/src/dirackernels/ReporterPointSource.C:
```diff
@@ -0,0 +1,88 @@
+ReporterPointSource::ReporterPointSource(const InputParameters & parameters)
+  : DiracKernel(parameters),
+    ReporterInterface(this),
+    _values(getReporterValueByName<std::vector<Real>>(
+        ReporterName(getParam<std::string>("value_name")), REPORTER_MODE_REPLICATED)),
+    _x_coord(getReporterValueByName<std::vector<Real>>(
+        ReporterName(getParam<std::string>("x_coord_name")), REPORTER_MODE_REPLICATED)),
+    _y_coord(getReporterValueByName<std::vector<Real>>(
+        ReporterName(getParam<std::string>("y_coord_name")), REPORTER_MODE_REPLICATED)),
+    _z_coord(getReporterValueByName<std::vector<Real>>(
+        ReporterName(getParam<std::string>("z_coord_name")), REPORTER_MODE_REPLICATED))
```
With `addRequiredParam<ReporterName>` you just have to do `_values(getReporterValue("value_name")),`

lynnmunday replied:
How can check that they are in REPORTER_MODE_REPLICATED

zachmprince replied:
Sorry, should be `_values(getReporterValue("value_name", REPORTER_MODE_REPLICATED)),

zachmprince commented on framework/src/dirackernels/ReporterPointSource.C:
```diff
@@ -0,0 +1,88 @@
+        ";  y_coord size = " + std::to_string(_y_coord.size()) +
+        ";  z_coord size = " + std::to_string(_z_coord.size());
+
+    mooseError(errMsg);
```
Maybe make this `paramError("value_name", errMsg)`

lynnmunday replied:
what is the difference?

zachmprince replied:
This will point the user to the line in the input file where `value_name` was specified.

lynnmunday commented:
@dschwen @zachmprince will you review this and merge?

zachmprince commented on framework/src/dirackernels/ReporterPointSource.C:
```diff
@@ -0,0 +1,84 @@
+        ";  y_coord size = " + std::to_string(_y_coord.size()) +
+        ";  z_coord size = " + std::to_string(_z_coord.size());
+
+    paramError(errMsg);
```
```suggestion
    paramError("value_name", errMsg);
```

lynnmunday replied:
I changed this back to a mooseError since I don't know which parameter is the wrong size.

zachmprince commented on framework/src/dirackernels/ReporterPointSource.C:
```diff
@@ -0,0 +1,84 @@
+ReporterPointSource::ReporterPointSource(const InputParameters & parameters)
+  : DiracKernel(parameters),
+    ReporterInterface(this),
+    _values(getReporterValue<std::vector<Real>>("value_name")),
+    _x_coord(getReporterValue<std::vector<Real>>("x_coord_name")),
+    _y_coord(getReporterValue<std::vector<Real>>("y_coord_name")),
+    _z_coord(getReporterValue<std::vector<Real>>("z_coord_name"))
```
```suggestion
    _values(getReporterValue<std::vector<Real>>("value_name", REPORTER_MODE_REPLICATED)),
    _x_coord(getReporterValue<std::vector<Real>>("x_coord_name", REPORTER_MODE_REPLICATED)),
    _y_coord(getReporterValue<std::vector<Real>>("y_coord_name", REPORTER_MODE_REPLICATED)),
    _z_coord(getReporterValue<std::vector<Real>>("z_coord_name", REPORTER_MODE_REPLICATED))
```

lynnmunday committed:
Update framework/src/dirackernels/ReporterPointSource.C

Co-authored-by: Zachary Prince <zachmprince@gmail.com>

lynnmunday commented:
@zachmprince @dschwen I address the comments, could you review it again

lynnmunday commented:
@dschwen will you check that I've resolved your issues and merge this?

dschwen reviewed (approved):
LGTM
